client/src/main/scala/com.crobox.clickhouse/stream/ClickhouseSink.scala

```diff
@@ -29,24 +44,55 @@ object ClickhouseSink extends LazyLogging {
           else None
       )
       .getOrElse(indexerGeneralConfig)
-    val batchSize     = mergedIndexerConfig.getInt("batch-size")
+    val batchSize = mergedIndexerConfig.getInt("batch-size")
     val flushInterval = mergedIndexerConfig.getDuration("flush-interval").getSeconds.seconds
-    Flow[Insert]
+    Flow[TableOperation]
       .groupBy(Int.MaxValue, _.table)
       .groupedWithin(batchSize, flushInterval)
-      .mapAsyncUnordered(mergedIndexerConfig.getInt("concurrent-requests"))(inserts => {
-        val table       = inserts.head.table
-        val insertQuery = s"INSERT INTO $table FORMAT JSONEachRow"
-        val payload     = inserts.map(_.jsonRow)
-        logger.debug(s"Inserting ${inserts.size} entries in table: $table. Group Within: ($batchSize - $flushInterval)")
-        client
-          .execute(insertQuery, payload.mkString("\n"))
-          .recover {
-            case ex => throw ClickhouseIndexingException("failed to index", ex, payload, table)
-          }
-          .map(_ => inserts)
+      .mapAsync(mergedIndexerConfig.getInt("concurrent-requests"))(operations => {
+        val table = operations.head.table
+        var insertPayload: Seq[String] = Seq.empty
+        var optimizeQuery: Option[String] = None
+
+        operations.foreach {
+          case o: Insert => insertPayload = insertPayload :+ o.jsonRow
+          case o: Optimize => optimizeQuery = Some(buildOptimizeQuery(o.table, o.cluster))
+        }
+
+        logger.debug(s"Inserting ${insertPayload.size} entries in table: $table. Group Within: ($batchSize - $flushInterval)")
+        insertOp(table, insertPayload, client)
+          .flatMap(_ => optimizeQuery match {
+            case Some(o) => optimizeOp(table, o, client)
+            case _ => Future.successful("")
+          })
+          .map(_ => operations)
       })
       .mergeSubstreams
       .toMat(Sink.ignore)(Keep.right)
   }
+
+  private def insertOp(table: String, payload: Seq[String], client: ClickhouseClient)(
```

Comment:
Please use a sequence of [InsertOperation' instead of json over here. Please maintain logic how to handle the different implementation (of the operation) in place